src/util/report-util.js

```diff
@@ -0,0 +1,305 @@
+    const dataForCSV = data.transactionDetails.map(tx => ({
+      'Time': tx.time,
+      'To': tx.to,
+      'Amount': tx.amount,
+      [`Amount in ${data.fiat}`]: tx.amountFiat,
+      'Balance': tx.balance,
+      [`Balance in ${data.fiat}`]: tx.balanceFiat,
+      'Description': tx.description
+    }))
+    const fields = ['Time', 'To', 'Amount', `Amount in ${data.fiat}`, 'Balance', `Balance in ${data.fiat}`, 'Description']
+
+    let csv
+
+    try {
+      csv = json2csv(dataForCSV, { fields })
+    } catch (err) {
+      return reject(err)
+    }
+
+    const blob = new Blob(
+      [csv],
+      { type: 'text/csv;charset=utf-8' }
+    )
+
+    debug('successfully generated CSV output', csv)
+
+    resolve({ filename: data.filename, blob })
+  })
+}
+
+/**
+ * Generate a report data object
+ * @param {Object} params
+ * @param {String} params.accountId - e.g. 'alice@test'
+ * @param {Object} params.wallet - { name, asset, precision, amount }
+ * @param {Object[]} params.transactins - an array of TransferAsset transactions
+ * @param {String} params.fiat - e.g. 'RUB'
+ * @param {Number} params.priceFiat - a price of fiat
+ * @param {Date} params.dateFrom - start of the range
+ * @param {Date} params.dateTo - end of the range
+ * @param {String} params.ext - e.g. 'pdf'
+ * @param {Function} params.formatDate - a functino to format a date to the specific string
+ * @param {Function} params.formatDateWith - a function to format a date to an arbitrary string
+ * @returns {Object}
+ */
+export function generateReportData ({
+  accountId,
+  wallet,
+  transactions,
+  priceFiat,
+  dateFrom,
+  dateTo,
+  ext,
+  formatDate,
+  formatDateWith,
+  fiat
+}) {
+  /*
+   * prepare utility functions
+   */
+  const sumAmount = (txs) => txs
+    .map(tx => parseFloat(tx.amount))
+    .reduce((sum, x) => sum + x, 0)
+  const isToYou = (tx) => (tx.to === 'you')
+  const isFromYou = (tx) => (tx.from === 'you')
+
+  /*
+   * prepare basic data
+   */
+  const filename = `report-${wallet.name.toLowerCase()}-${formatDateWith(dateFrom, 'YYYYMMDD')}-${formatDateWith(dateTo, 'YYYYMMDD')}.${ext}`
+  const cryptoCurrencyUnit = wallet.asset
+  const precision = wallet.precision
+  const currentAmount = wallet.amount
+
+  /*
+   * split transactions
+   */
+  const txsWithinRange = transactions.filter(tx => isWithinRange(tx.date, startOfDay(dateFrom), endOfDay(dateTo)))
+  const txsAfterRange = transactions.filter(tx => isAfter(tx.date, endOfDay(dateTo)))
+
+  /*
+   * calculate figures
+   */
+  const netChangeAfterRange = txsAfterRange
+    .map(tx => (tx.to === 'you') ? +tx.amount : -tx.amount)
+    .reduce((sum, x) => sum + x, 0)
+  const transfersIn = sumAmount(txsWithinRange.filter(isToYou))
+  const transfersOut = sumAmount(txsWithinRange.filter(isFromYou))
+  const netChange = transfersIn - transfersOut
+  const endingBalance = currentAmount - netChangeAfterRange
+  const endingBalanceFiat = endingBalance * priceFiat
```

Comment:
You should not multiply `priceFiat`, because key `price` in `portfolioList` don't mean price of this currency, it's mean full price (priceForDay * amount). Sorry for unclear key.

Or another way is to use `portfolioHistory`. It's array of prices for each wallet for one month:
```javascript
const priceFiat = this.portfolioHistory[this.portfolioHistory.length - 1]
  .data.find(({ asset }) => asset === wallet.asset).value.close
```
It will give you price for one day.